File: samples/cpp/ods_header.cpp
/*
 *  ods_header.cpp — companion to ../../on-disk-structure.md
 *
 *  Reads a database file's header page (page 0) directly from disk and
 *  decodes it with the byte offsets that src/jrd/ods.h pins with
 *  static_asserts: page type, page size, ODS version, the hdr_PAGES
 *  bootstrap anchor, and the four transaction markers (Next/OIT/OAT/OST).
 *  Then it walks the whole file reading byte 0 of every page — the page
 *  type — to show "one file, many pages" as a census.
 *
 *  Steps: create/attach a scratch database through the server, ask
 *  MON$DATABASE for the server's own view of the same numbers, detach,
 *  then open the file and parse raw bytes.  Run it from the machine the
 *  server runs on (the server writes /tmp/fbhandson/ods.fdb; we read it).
 */

#include "fb_sample.h"
#include <cstdint>

using namespace fbsample;

static uint16_t u16(const unsigned char* b, size_t o) { uint16_t v; memcpy(&v, b + o, 2); return v; }
static uint32_t u32(const unsigned char* b, size_t o) { uint32_t v; memcpy(&v, b + o, 4); return v; }
static uint64_t u64(const unsigned char* b, size_t o) { uint64_t v; memcpy(&v, b + o, 8); return v; }

static const char* pageTypeName(int t)
{
	static const char* names[] = { "undefined", "pag_header", "pag_pages (PIP)",
		"pag_transactions (TIP)", "pag_pointer", "pag_data", "pag_root",
		"pag_index (b-tree)", "pag_blob", "pag_ids (generators)", "pag_scns" };
	return (t >= 0 && t <= 10) ? names[t] : "???";
}

int main(int argc, char** argv)
{
	const char* database = argOrDefault(argc, argv, 1, "inet://localhost//tmp/fbhandson/ods.fdb");
	const char* localFile = argOrDefault(argc, argv, 2, "/tmp/fbhandson/ods.fdb");

	try
	{
		// 1. Create (or reuse) the scratch database and generate a little
		//    transaction history so the TIP markers move off their floor.
		Db db;
		db.attachOrCreate(database);
		for (int i = 0; i < 3; ++i)
		{
			ITransaction* t = db.start();
			db.queryValue(t, "select 1 from rdb$database");
			t->commit(&db.status);
		}

		ITransaction* tra = db.start();
		printf("-- server's view (MON$DATABASE) --\n");
		Db::print(db.query(tra,
			"select mon$page_size, mon$ods_major, mon$ods_minor,"
			"       mon$oldest_transaction, mon$oldest_active,"
			"       mon$oldest_snapshot, mon$next_transaction "
			"from mon$database"));
		tra->commit(&db.status);
		db.att->detach(&db.status);
		db.att = nullptr;

		// 2. Now the same facts straight from the bytes on disk.
		FILE* f = fopen(localFile, "rb");
		if (!f)
			{ perror(localFile); return 1; }

		unsigned char h[152];					// sizeof(Ods::header_page)
		if (fread(h, 1, sizeof h, f) != sizeof h)
			{ fprintf(stderr, "short read\n"); return 1; }

		printf("\n-- header page, parsed from %s (offsets per ods.h) --\n", localFile);
		printf("pag_type      @0   = %u (%s)\n", h[0], pageTypeName(h[0]));
		printf("pag_flags     @1   = %u\n", h[1]);

		const unsigned pageSize = u16(h, 16);
		const uint16_t odsRaw = u16(h, 18);
		printf("hdr_page_size @16  = %u\n", pageSize);
		printf("hdr_ods_version @18 = 0x%04x -> ODS %u (FIREBIRD flag 0x8000 %s), minor @20 = %u\n",
			odsRaw, odsRaw & 0x7fff, (odsRaw & 0x8000) ? "set" : "clear", u16(h, 20));

		const uint16_t flags = u16(h, 22);
		printf("hdr_flags     @22  = 0x%02x (%s%s%s)\n", flags,
			(flags & 0x2) ? "force_write " : "", (flags & 0x8) ? "no_reserve " : "",
			(flags & 0x10) ? "SQL_dialect_3" : "");
		printf("hdr_PAGES     @28  = %u   <- pointer page of RDB$PAGES (catalog bootstrap anchor)\n",
			u32(h, 28));
		printf("hdr_next_transaction   @40 = %llu\n", (unsigned long long) u64(h, 40));
		printf("hdr_oldest_transaction @48 = %llu (OIT)\n", (unsigned long long) u64(h, 48));
		printf("hdr_oldest_active      @56 = %llu (OAT)\n", (unsigned long long) u64(h, 56));
		printf("hdr_oldest_snapshot    @64 = %llu (OST)\n", (unsigned long long) u64(h, 64));

		const unsigned char* g = h + 84;		// hdr_guid: Win32 GUID layout
		printf("hdr_guid      @84  = {%08X-%04X-%04X-%02X%02X-%02X%02X%02X%02X%02X%02X}\n",
			u32(g, 0), u16(g, 4), u16(g, 6), g[8], g[9], g[10], g[11], g[12], g[13], g[14], g[15]);

		// 3. Page-type census: byte 0 of every page in the file.
		unsigned counts[11] = {};
		unsigned long pageNo = 0;
		for (;; ++pageNo)
		{
			if (fseek(f, (long) (pageNo * (unsigned long) pageSize), SEEK_SET) != 0)
				break;
			int c = fgetc(f);
			if (c == EOF)
				break;
			counts[(c >= 0 && c <= 10) ? c : 0]++;
		}
		fclose(f);

		printf("\n-- page-type census: %lu pages of %u bytes --\n", pageNo, pageSize);
		for (int t = 1; t <= 10; ++t)
			if (counts[t])
				printf("  type %2d  %-22s %5u\n", t, pageTypeName(t), counts[t]);
		printf("done.\n");
		return 0;
	}
	catch (const FbException& error)
	{
		return report(error);
	}
}

File: samples/nodejs/ods_header.js
//
// ods_header.js — companion to ../../on-disk-structure.md
//
// The JavaScript twin of ../cpp/ods_header.cpp, and a natural fit for Node:
// Buffer.readUInt16LE / readBigUInt64LE parse the header page (page 0)
// directly at the byte offsets src/jrd/ods.h pins with static_asserts.
// Creates a scratch database through the server, compares MON$DATABASE's
// view with the raw bytes, then runs a page-type census over the file.
// Run on the server machine: node ods_header.js
//
'use strict';

const fs = require('fs');
const { attachOrCreate, s } = require('./common');

const DB_REMOTE = '/tmp/fbhandson/ods_js.fdb';
const DB_LOCAL = '/tmp/fbhandson/ods_js.fdb';

const PAGE_TYPES = ['?', 'pag_header', 'pag_pages (PIP)', 'pag_transactions (TIP)',
    'pag_pointer', 'pag_data', 'pag_root', 'pag_index (b-tree)', 'pag_blob',
    'pag_ids (generators)', 'pag_scns'];

(async () => {
    // 1. Create/attach through the server; a few commits move the TIP markers.
    const conn = await attachOrCreate({ database: DB_REMOTE });
    for (let i = 0; i < 3; i++)
        await conn.query('select 1 from rdb$database');

    const [mon] = await conn.query(
        `select mon$page_size ps, mon$ods_major omaj, mon$ods_minor omin,
                mon$oldest_transaction oit, mon$oldest_active oat,
                mon$oldest_snapshot ost, mon$next_transaction nxt
           from mon$database`);
    console.log("-- server's view (MON$DATABASE) --");
    console.log(`page size ${mon.PS}, ODS ${mon.OMAJ}.${mon.OMIN},` +
        ` OIT ${mon.OIT}, OAT ${mon.OAT}, OST ${mon.OST}, next ${mon.NXT}`);
    await conn.detach();

    // 2. The same facts from the bytes.  struct header_page is 152 bytes.
    const file = fs.readFileSync(DB_LOCAL);        // whole file; census needs it anyway
    const h = file.subarray(0, 152);

    const odsRaw = h.readUInt16LE(18);
    const pageSize = h.readUInt16LE(16);
    const guid = (b) => (`{${b.readUInt32LE(0).toString(16).padStart(8, '0')}-` +
        `${b.readUInt16LE(4).toString(16).padStart(4, '0')}-` +
        `${b.readUInt16LE(6).toString(16).padStart(4, '0')}-` +
        `${b.subarray(8, 10).toString('hex')}-` +
        `${b.subarray(10, 16).toString('hex')}}`).toUpperCase();

    console.log(`\n-- header page, parsed from ${DB_LOCAL} (offsets per ods.h) --`);
    console.log(`pag_type      @0   = ${h[0]} (${PAGE_TYPES[h[0]]})`);
    console.log(`hdr_page_size @16  = ${pageSize}`);
    console.log(`hdr_ods_version @18 = 0x${odsRaw.toString(16)} -> ODS ${odsRaw & 0x7fff}` +
        ` (FIREBIRD flag ${odsRaw & 0x8000 ? 'set' : 'clear'}), minor @20 = ${h.readUInt16LE(20)}`);
    const flags = h.readUInt16LE(22);
    console.log(`hdr_flags     @22  = 0x${flags.toString(16)}` +
        `${flags & 0x2 ? ' force_write' : ''}${flags & 0x10 ? ' SQL_dialect_3' : ''}`);
    console.log(`hdr_PAGES     @28  = ${h.readUInt32LE(28)}` +
        '   <- pointer page of RDB$PAGES (catalog bootstrap anchor)');
    console.log(`hdr_next_transaction   @40 = ${h.readBigUInt64LE(40)}`);
    console.log(`hdr_oldest_transaction @48 = ${h.readBigUInt64LE(48)} (OIT)`);
    console.log(`hdr_oldest_active      @56 = ${h.readBigUInt64LE(56)} (OAT)`);
    console.log(`hdr_oldest_snapshot    @64 = ${h.readBigUInt64LE(64)} (OST)`);
    console.log(`hdr_guid      @84  = ${guid(h.subarray(84, 100))}`);

    // 3. Page-type census: byte 0 of every page.
    const counts = new Map();
    for (let off = 0; off < file.length; off += pageSize)
        counts.set(file[off], (counts.get(file[off]) || 0) + 1);
    console.log(`\n-- page-type census: ${file.length / pageSize} pages of ${pageSize} bytes --`);
    for (const t of [...counts.keys()].sort((a, b) => a - b))
        console.log(`  type ${String(t).padStart(2)}  ${PAGE_TYPES[t].padEnd(22)} ${String(counts.get(t)).padStart(5)}`);
    console.log('done.');
})().catch(e => { console.error('ERR:', e.message); process.exit(1); });
